lib/hypgen/planner.rb

```diff
@@ -1,25 +1,64 @@
 require 'json'
 require 'recursive-open-struct'
+require 'date'
 
 module Hypgen
   class Planner
 
+    # number of days in time window, rounded down to the full days
+    attr_reader :days
+
     def initialize(workflow)
       @workflow = workflow
       @deadline = @workflow.deadline
-      @constant_term = 40
+      start_date = DateTime.parse(@workflow.start_time)
+      end_date = DateTime.parse(@workflow.end_time)
+      @days = (end_date - start_date).to_i
+      @days = 1 if days <= 0
     end
 
-    def estimate_task_runtime
-      # return current estimation from the experiments
-      9.345820862
+    def estimate_vm_count(wf)
+      s = count_processes(wf)
+      d = @days
+      t = @deadline
+
+      vm_count = compute_perf_model(s, d, t)
+      return vm_count
     end
 
-    def estimate_vm_count(wf)
-      vmc = count_processes(wf) * estimate_task_runtime / ( @deadline - @constant_term)
+    # Our performance model
+    # Model Parameters:
+    # T - time (deadline) in seconds
+    # s - number of sections
+    # d - number of days (window size)
+    # v - number of VMs
+    #
+    # Function parameters obtained from fit to data:
+    # a = 6.53
+    # b = 9.41
+    # c = 31.71
+    #
+    # Function:
+    # T = a * s * d / v + b * v + c
+    #
+    # This can be solved to find v:
+    # b * v^2 + (c-T) * v + a * s * d = 0
+    # This gives:
+    #
+    # v = (-(c-T) +/- sqrt((c-T)^2 -4 * b * a * s *d)) / (2 * b)
+    # Out of two solutions we select the smaller one, so:
+    # v = (-(c-T) - sqrt((c-T)^2 - 4 * b * a * s *d)) / (2 * b)
+
+    def compute_perf_model(s, d, t)
```

Comment:
Well, this is a performance model, it must be complex :)

On 04/01/2015 05:07 PM, Hound wrote:
>
> In lib/hypgen/planner.rb 
> <https://github.com/ismop/hypgen/pull/10#discussion_r27577726>:
>
> > +    # a = 6.53
> > +    # b = 9.41
> > +    # c = 31.71
> > +    #
> > +    # Function:
> > +    # T = a * s * d / v + b * v + c
> > +    #
> > +    # This can be solved to find v:
> > +    # b * v^2 + (c-T) * v + a * s * d = 0
> > +    # This gives:
> > +    #
> > +    # v = (-(c-T) +/- sqrt((c-T)^2 -4 * b * a * s *d)) / (2 * b)
> > +    # Out of two solutions we select the smaller one, so:
> > +    # v = (-(c-T) - sqrt((c-T)^2 - 4 * b * a * s *d)) / (2 * b)
> > +
> > +    def compute_perf_model(s, d, t)
>
> Assignment Branch Condition size for compute_perf_model is too high. 
> [18.49/15]
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/ismop/hypgen/pull/10/files#r27577726>.
>